gloss/Graphics/Gloss/Geometry/Triangulation.hs

```diff
@@ -0,0 +1,388 @@
+import Graphics.Gloss.Rendering
+import Graphics.Gloss.Geometry.Line
+import Graphics.Gloss.Geometry.Angle
+import Graphics.Gloss.Data.Vector
+import qualified Data.Set as Set
+import qualified Data.Map.Strict as Map
+import Data.List (minimumBy, maximumBy, sortBy)
+import Control.Applicative
+import Data.Maybe
+
+-- import Debug.Trace
+
+
+-- | Type for representing planar graphs.
+type Vertices = Map.Map Vertex (Set.Set Vertex)
+
+-- | Whrepper for @Point@. Has different @Eq@ implmentation to absorb numerical error in calculations.
+data Vertex
+        = Vertex Point
+        deriving (Show)
+
+-- | Utility function to convert @Vertex@ to @Point@
+toPoint :: Vertex -> Point
+
+toPoint (Vertex x) = x
+
+instance Eq Vertex where
+        (Vertex (x1,y1)) == (Vertex (x2,y2)) = abs (x1-x2) + abs (y1-y2) < 0.0001
+
+instance Ord Vertex where
+        compare v1@(Vertex (x1,y1)) v2@(Vertex (x2,y2))
+                | v1 == v2 = EQ
+                | otherwise = compare (x1,y1) (x2,y2)
+
+data Edge
+        = Edge
+        { start :: Vertex
+        , end :: Vertex }
+        deriving (Show)
+
+instance Eq Edge where
+        Edge{start = p0, end = p1} == Edge{start = p2, end = p3}
+                = p0 == p2 && p1 == p3
+instance Ord Edge where
+        Edge{start = Vertex p0@(x0, _), end = Vertex p1@(x1, y1)} < Edge{start = Vertex p2@(x2, _), end = Vertex p3@(x3, y3)}
+                            = let x = max x0 x2
+                                  intersectHelper a@(ax,_) b@(bx,_) yAxisXCoorcinate
+                                        = if ax == bx && ax == yAxisXCoorcinate then Just a else intersectSegVertLine a b yAxisXCoorcinate
+                                  Just (_, py) = intersectHelper p0 p1 x
+                                  Just (_, qy) = intersectHelper p2 p3 x
+                              in if py == qy then (x1-x)*(y3-py) > (x3-x)*(y1-py) else py < qy
+        compare x y
+                | x == y = EQ
+                | x < y = LT
+                | otherwise = GT
+
+
+-- | Map of events for processing. Holds segments that end in given $Vertex$ and segments that begin there.
+type Events = Map.Map Vertex (Set.Set Edge, Set.Set Edge)
+
+-- | Determen if points are close enough.
+(!=) :: Point -> Point -> Bool
+
+(!=) p1 p2 = Vertex p1 /= Vertex p2
+
+-- | Finds intersections of two Edges.
+intersection :: Maybe Edge -> Maybe Edge -> Maybe Vertex
+
+intersection (Just Edge{start = Vertex p0, end = Vertex p1}) (Just Edge{start = Vertex p2, end = Vertex p3})
+                | Just x <- intersectSegSeg p0 p1 p2 p3
+                , (x != p0) && (x != p1) && (x != p2) && ( x != p3)
+                = Just (Vertex x)
+                | otherwise = Nothing
+intersection _ _ = Nothing
+
+-- | Turns @Path@ into a graph in witch egdes can intersect. You can think of it as a drawing of a @Path@.
+makeInitialVertexSet :: Path -> Vertices
+
+makeInitialVertexSet [] = undefined
```

Comment:
We can't have uses of 'undefined' in this code, as if the user ever causes this to be evaluated there will be no description of what went wrong. If it should not be possible to cause this error via the public API then use the error function with a message describing what went wrong in the implementation.